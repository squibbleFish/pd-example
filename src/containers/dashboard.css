.theme-board {
  width: auto;
  display: flex;
  flex-flow: row wrap;
  justify-content: space-evenly;
  align-items: baseline;
}

.theme-card {
  order: 3;
  width: 33%;
  border: 1px solid #d9d9dc;
  border-radius: 2px;
  padding-bottom: 1rem;
  padding-top: 1rem;
  text-align: center;
}

.spinner {
  border: 16px solid #f7f7f9;
  border-top: 16px solid #292b2c;
  border-radius: 50%;
  width: 120px;
  height: 120px;
  animation: spin 2s linear infinite;
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}

.btn {
  display: inline-block;
  text-decoration: none;
  color: #fff;
  border: 1px solid #000;
  padding: 10px;
  font-size: 12px;
  background-color: #c3a312;
}

span.theme-type._default {
  background: #c3a312;
  color: #fff;
  border-radius: 9px;
  margin: 10px;
  padding: 5px;
}

span.theme-type._default:after {
  content: "Default";
  font-size: 12px;
}