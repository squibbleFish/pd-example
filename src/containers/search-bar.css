.search-bar {
    order: 1;
    width: 100%;
    padding: 10px 20px;
    margin: 0 auto;
}

input.search {
    padding: 13px;
    width: 100%;
    border: 1px solid;
    border-radius: 4px;
}