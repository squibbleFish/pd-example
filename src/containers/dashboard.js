import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import _ from 'lodash';
import { Link, withRouter } from 'react-router-dom';

import * as actions from '../actions/theme-actions';
import SearchBar from './search-bar';
import { Spinner } from '../components/spinner';
import './dashboard.css';

export class Dashboard extends Component {
  /**
   *
   * @param props
   */
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      error: false,
      displayed: [],
    };
  }

  /**
   *
   * @returns {Promise<void>}
   */
  componentWillMount() {
    this.props.getThemes();
  }

  /**
   * Check to make sure our data is back
   * @param nextProps
   */
  componentWillReceiveProps(nextProps) {
    /**
     * Loader and initial theme load
     */
    if (!_.isEmpty(nextProps.themes.themes)) {
      this.setState({
        loading: false,
        displayed: nextProps.themes.themes
      })
    }

    /**
     * handles the search functionality
     */
    if(!_.isNil(nextProps.themes.search)) {
      const term = nextProps.themes.search;
      let filtered = _.filter(this.props.themes.themes, t => _.toLower(t.title).includes(term.toLowerCase()));
      this.setState({
        displayed: filtered
      });
    }

    /**
     *
     */
    if (nextProps.themes.error) {
      this.setState({
        error: true
      }, () => {
        // We could throw to a log on the server if need be
        console.log(nextProps.themes.trace);
      })
    }
  }

  /**
   *
   * @returns {*}
   */
  render() {
    const themes = !_.isEmpty(this.state.displayed) ? this.state.displayed : [];
    /**
     * Map the themes
     * @type {string}
     */
    const dash = !_.isEmpty(themes) ? themes.map((theme) => {
      const owner = theme.owner === '_default' ? '' : theme.owner;
      return(
        <div
          className="theme-card"
          key={ theme.guid } >
          <h4>
            { theme.title }
            <span className={ `theme-type ${ theme.owner }`} />
          </h4>
          <p>{ !_.isEmpty(owner) ? `Owner : ${ owner }` : '' }</p>
          <Link
            to={{
              pathname: `/theme/${ theme.guid }`,
              state: { theme }
            }}
            className="btn preview" >
            Preview Theme
          </Link>
        </div>
      )
    }) : '';

    return(
      <React.Fragment>
        {
          this.state.error ?
            <div className="error">There was an error. Please try again</div> : ''
        }
        <div
          className="theme-board" >
          <SearchBar />
          {
            // Either of these could be components, but this is just slightly easier
            this.state.loading ? <Spinner /> : dash
          }
        </div>
      </React.Fragment>
    )
  }
}

function mapStateToProps(state) {
  return {
    themes: state.themes
  }
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(actions, dispatch);
}

Dashboard.propTypes = {
  getThemes: PropTypes.func,
  searchThemes: PropTypes.func,
  themes: PropTypes.object,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps, null, { withRef: true})(Dashboard));