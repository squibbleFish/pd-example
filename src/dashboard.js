import React, { Component } from 'react';
import Request from 'axios';
import _ from 'lodash';
import { Link } from 'react-router-dom';

import { BASE } from './config';
import './dashboard.css';

/**
 * API URL
 * @type {string}
 */
const API = `${BASE}/api/theme`;
/**
 * API token
 * @type {string}
 */
const TOKEN = '15151ca0-ffbf-4c13-ac7d-89b5e89a417c';

class Dashboard extends Component {
  /**
   *
   * @param props
   */
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      error: false,
      themes: {}
    };
  }

  /**
   * Async function to grab the dataset.
   * @returns {Promise<void>}
   */
  async componentDidMount() {
    // Fetch data from the API
    try {
      const { data } = await Request.get(API, {
        headers: { 'Authorization': `Bearer ${ TOKEN }` }
      });
      await this.setThemes({themes: data.rows});
      this.setState({
        loading: false
      });
    } catch (e)  {
      console.log(e); // The stacktrace could be sent to a log on the server if need be
      this.setState({
        error: true
      });
    }
  }

  /**
   * Once themes have been returned set them in the state
   * @param state
   * @returns {Promise<any>}
   */
  setThemes(state) {
    return new Promise((resolve) => {
      this.setState(state, resolve)
    });
  }

  /**
   *
   * @returns {*}
   */
  render() {
    /**
     * Map the themes
     * @type {string}
     */
    const themes = !_.isEmpty(this.state.themes) ? this.state.themes.map((theme, i) => {
      const owner = theme.owner === '_default' ? '' : theme.owner;
      return(
        <div
          className="theme-card"
          key={ theme.guid } >
          {
            this.state.error ?
              <div className="error">There was an error. Please try again</div> : ''
          }
          <h4>
            { theme.title }
            <span className={ `theme-type ${theme.owner}`} />
          </h4>
          <p>
            {
              !_.isEmpty(owner) ? `Owner : ${owner}` : ''
            }
          </p>
          <Link
            to={{
              pathname: `/theme/${theme.guid}`,
              state: { theme }
            }}
            className="btn preview" >
            Preview Theme
          </Link>
        </div>
      )
    }) : '';
    const spinner = <div className="spinner" />;
    return(
      <div className="theme-board">
        {
          this.state.loading ? spinner : themes
        }
      </div>
    )
  }
}

export default Dashboard;