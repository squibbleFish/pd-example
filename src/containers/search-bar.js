import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';

import * as actions from '../actions/search-actions';
import './search-bar.css';

class SearchBar extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }


  handleChange(e) {
    this.props.searchThemes(e.target.value);
  }

  render() {
    return(
      <section
        className="search-bar">
        <input
          type="search"
          name="search"
          className="search"
          placeholder="Enter Your Search Term"
          value={ this.props.search.search }
          onChange={ this.handleChange } />
        <a
          href="#" >
        </a>
      </section>
    );
  }
}

function mapStateToProps(state) {
  return {
    search: state.search
  }
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(actions, dispatch);
}

SearchBar.propTypes = {
  searchThemes: PropTypes.func,
  search: PropTypes.object,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps, null, { withRef: true })(SearchBar))