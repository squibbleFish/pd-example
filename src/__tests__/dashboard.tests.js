import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk';

import { Dashboard } from '../containers/dashboard';

const mockedJson = {
  themes: [{
    owner:"_default",
    handle:"sidekick",
    type:"theme",
    guid:"e11ca530-fd63-11e7-a5be-e53eb9f02238",
    created:"2018-01-19T21:58:28.528Z",
    last_updated:"2018-01-19T21:58:31.792Z",
    title:"Sidekick",
    description:"A theme ready for action, with a hero cover image and a sidebar.",
    main:"73ac0428-8400-4202-abde-7ca1bf46ad09",
  }]
};

const props = {
  getThemes: jest.fn(),
  themes: { themes: { mockedJson } },
  search: { search: '' },
};

describe('', () => {
  it('should render', function () {
    const wrapper = shallow(
      <Dashboard { ...props } />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a box to contain a single theme', function () {
    const mockStore = configureMockStore([thunk]);
    const store = mockStore({ search: { search: '' }, themes: { themes: mockedJson } });
    const wrapper = mount(
      <Provider store={store} >
        <MemoryRouter>
          <Dashboard { ...props } />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find('.theme-card')).toHaveLength(1);
  });
});
