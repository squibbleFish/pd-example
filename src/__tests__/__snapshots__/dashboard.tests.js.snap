// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dashboard
    getThemes={[Function]}
    search={
        Object {
            "search": "",
          }
    }
    themes={
        Object {
            "themes": Object {
              "mockedJson": Object {
                "themes": Array [
                  Object {
                    "created": "2018-01-19T21:58:28.528Z",
                    "description": "A theme ready for action, with a hero cover image and a sidebar.",
                    "guid": "e11ca530-fd63-11e7-a5be-e53eb9f02238",
                    "handle": "sidekick",
                    "last_updated": "2018-01-19T21:58:31.792Z",
                    "main": "73ac0428-8400-4202-abde-7ca1bf46ad09",
                    "owner": "_default",
                    "title": "Sidekick",
                    "type": "theme",
                  },
                ],
              },
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        "",
        <div
          className="theme-board"
>
          <withRouter(Connect(SearchBar)) />
          <Unknown />
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <withRouter(Connect(SearchBar)) />,
            <Unknown />,
          ],
          "className": "theme-board",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          "",
          <div
            className="theme-board"
>
            <withRouter(Connect(SearchBar)) />
            <Unknown />
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <withRouter(Connect(SearchBar)) />,
              <Unknown />,
            ],
            "className": "theme-board",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
